### ISSUE TITLE
Blocked user can see your toot and talk in your toot's comments without your knowledge if you mention them

### ISSUE THREAD

**Issue by Kaukamieli** [mastodon/mastodon#19698]
### Steps to reproduce the problem

Have your own server or a friend to test so you see both accounts.
1.Block a user
2.Mention blocked user in a "private message". e.g. "My mom @account@duh is abusing me. I blocked them so they can't see this..."
3.As blocked user read the message, reply to it, and talk with the whoever the blocker was talking with without blocker knowing anything.

### Expected behaviour

Blocked user should never see your toots or be able to reply to them, because users will mention them and recommend to block and talk about their behavior, and have no idea how block actually works.

### Actual behaviour

Blocked user sees all messages they are tagged in, even when blocked, and can reply and talk in your thread without you knowing

### Detailed description

Block is a security risk and risk to users.

Users would expect blocker to not be able to see any of your posts or comment on your posts.

Currently if you block a person, they:
1. Get notified when you mention them.
2. See who you are talking with, because you mention to answer.
3. Can answer your toot, because block hides instead of blocking.
4. Can talk with your friends in your thread, because block does not prevent commenting if they see it. And they see it because you mentioned it. And you do not see any of it because you blocked them, so you do not see it when they comment, or when someone replies to them.

Scenario 1: Your abusive mom

Your mother is abusive and is in Mastodon.
You block them and warn your friend with mention only. But you used full handle of your mom because you blocked them, so they should not see anything, right? Wrong.

They get notified and can comment to your toot.

You do not see what they comment, and can not see what your friends reply to them. You think you are having a personal discussion about your mom, but she is lying about you to your closest friends. And you do not know unless your friend tells you.

Scenario 2: Nazi

You block a nazi and warn people there are nazis around.

Someone asks you in private about what the account is so they can block them.

You answer mentioning the whole handle in private, or you think so because you blocked them.

Nazi sees this, and what you talk about them, and with whom.

They get offended and can attack you physically if they have access for talking shit about them or anything

***********

Seriously please fix blocking please. There should probably be no reason blocked people get any information of you talking about them and they should not be able to comment on your stuff. And they definitely should not be able to talk about you without you knowing when they know instantly if you who blocked them talk about them!

### Specifications

Web version running in instance c.im, so v3.5.3 at the moment.

**Comment by ClearlyClaire** [comment]
Hi, thanks for your report!

In general, while blocks should ensure you never see content from the blocked account, preventing them from seeing your content is a best-effort mechanism that cannot always be enforced with 100% certainty across instance boundaries.

That being said, the case you are describing sounds like a very clear oversight: I see no reason why mentions to a blocked account should be processed, and processing them clearly opens to damageable accidental mentions as you are describing. I will try to address this specific case as soon as possible.

**Comment by ClearlyClaire** [comment]
I believe I have addressed the “blocked people see and are notified of messages they are mentioned in” part in #19725, so I have marked it as fixing the issue, but keep in mind that even with that, blocks remain a best-effort mechanism and blocked people on servers that are not aware of, or ignore blocks, can reply to your posts.

There are some efforts to improve that situation, with for instance #18468 and #14762, but those are more involved changes that will require time and collaboration with other fediverse projects.

**Comment by Kaukamieli** [comment]
Thank you. Could they be denied from commenting to blocker, though too? If
some bug would allow them to see it, it would be nice to not let them
comment, as blocker can not see it and it would lead to this discussion
behind your back thing.

I agree that not processing the blocked mentions should be good enough to
close this if I understood things correctly, which I most probably did not.
Thank you for being quick, this felt huge.

**Comment by ClearlyClaire** [comment]
> Could they be denied from commenting to blocker, though too?

The best we can currently do is not processing their reply (this is roughly the road followed by https://github.com/mastodon/mastodon/pull/18468), but this would not prevent other servers from potentially seeing their reply, and those servers would not know that you have blocked that user. This is something that could be possible with stuff like https://github.com/mastodon/mastodon/issues/14762, though, but as I said previously, this is a significant amount of work which also requires coordination with other fediverse projects.

>  agree that not processing the blocked mentions should be good enough to close this if I understood things correctly, which I most probably did not.

To sum up the fix again: it fixes that blocked users are notified of posts in which you mention them, and it also fixes them being able to see the post when the only reason they can see the post is because they were notified of it.
